### ISSUE TITLE
To implement support for `tls-crypt`, new in OpenVPN v2.4.

### ISSUE THREAD

**Issue by Joyphor** [iphoting/ovpnmcgen.rb#15]
Hello,
I'm new to OpenVPN. I installed and configured PiVPN and found your tool to generate VoD-Profiles.
I have a **working** non-on-Demand config on iPhone with the OpenVPN Connect App.

PiVPN runs on a Raspberry Pi 3 (Raspbian Stretch) which acts as a DNS Server with pi-hole and DHCP behind a Fritzbox. Port forwarding is activ for port extern 31194 to intern 1194.
If I flip the switch in settings it says "Connecting..." until time out, in OpenVPN Connect it shows "Waiting for server". There are no log entries.

Following your readme, these are my settings:

**.ovpnmcgen.rb.yml**
```
#untrusted_ssids: [dangerous1, dangerous2]
trusted_ssids: [SSID1, SSID2]
host: home.xxxxxx.de
proto: UDP
port: 31194
domains: ["fritz.box", "*.fritz.box", "*.local"]
#cafile: /etc/openvpn/ca.crt
#tafile: /etc/openvpn/ta.key
#url_probe: https://vpn.example.com/canVPN.php
```

**iPhone.ovpn**
```
client
dev tun
proto udp
remote home.xxxxxx.de 31194
resolv-retry infinite
nobind
persist-key
persist-tun
remote-cert-tls server
tls-version-min 1.2
verify-x509-name server_avUhgcEDVcdST9MU name
cipher AES-256-CBC
auth SHA256
compress lz4
verb 3
```

**command**
`ovpnmcgen.rb gen --v12compat --cafile ./ovpns/iPhone-ca.crt --p12file ./ovpns/iPhone-client.p12 --p12pass xxxxxx --ovpnconfigfile ./ovpns/iPhone_nocert.ovpn --output iPhone.mobileconfig xxxxxx iPhone`

**output**
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>PayloadContent</key>
	<array>
		<dict>
			<key>PayloadDescription</key>
			<string>Configures VPN settings, including authentication.</string>
			<key>PayloadDisplayName</key>
			<string>VPN (home.xxxxxx.de/VoD)</string>
			<key>PayloadIdentifier</key>
			<string>de.xxxxxx.home.iPhone.vpnconfig</string>
			<key>PayloadOrganization</key>
			<string>home.xxxxxx.de</string>
			<key>PayloadType</key>
			<string>com.apple.vpn.managed</string>
			<key>PayloadUUID</key>
			<string>xxxxxx</string>
			<key>PayloadVersion</key>
			<integer>1</integer>
			<key>UserDefinedName</key>
			<string>home.xxxxxx.de/VoD</string>
			<key>VPN</key>
			<dict>
				<key>AuthenticationMethod</key>
				<string>Certificate</string>
				<key>OnDemandEnabled</key>
				<integer>1</integer>
				<key>OnDemandRules</key>
				<array>
					<dict>
						<key>Action</key>
						<string>Disconnect</string>
						<key>InterfaceTypeMatch</key>
						<string>WiFi</string>
						<key>SSIDMatch</key>
						<array>
							<string>SSID1</string>
							<string>SSID2</string>
						</array>
					</dict>
					<dict>
						<key>Action</key>
						<string>Connect</string>
						<key>InterfaceTypeMatch</key>
						<string>WiFi</string>
					</dict>
					<dict>
						<key>Action</key>
						<string>EvaluateConnection</string>
						<key>ActionParameters</key>
						<array>
							<dict>
								<key>DomainAction</key>
								<string>ConnectIfNeeded</string>
								<key>Domains</key>
								<array>
									<string>fritz.box</string>
									<string>*.fritz.box</string>
									<string>*.local</string>
								</array>
							</dict>
						</array>
					</dict>
					<dict>
						<key>Action</key>
						<string>Ignore</string>
						<key>InterfaceTypeMatch</key>
						<string>Cellular</string>
					</dict>
					<dict>
						<key>Action</key>
						<string>Ignore</string>
					</dict>
				</array>
				<key>PayloadCertificateUUID</key>
				<string>xxxxxx</string>
				<key>RemoteAddress</key>
				<string>DEFAULT</string>
			</dict>
			<key>VPNSubType</key>
			<string>net.openvpn.connect.app</string>
			<key>VPNType</key>
			<string>VPN</string>
			<key>VendorConfig</key>
			<dict>
				<key>auth</key>
				<string>SHA256</string>
				<key>ca</key>
				<string>-----BEGIN CERTIFICATE-----xxxxxx-----END CERTIFICATE-----</string>
				<key>cipher</key>
				<string>AES-256-CBC</string>
				<key>client</key>
				<string>NOARGS</string>
				<key>compress</key>
				<string>lz4</string>
				<key>dev</key>
				<string>tun</string>
				<key>remote</key>
				<string>home.xxxxxx.de 31194 UDP</string>
				<key>remote-cert-tls</key>
				<string>server</string>
				<key>tls-version-min</key>
				<string>1.2</string>
				<key>verify-x509-name</key>
				<string>server_avUhgcEDVcdST9MU name</string>
			</dict>
		</dict>
		<dict>
			<key>Password</key>
			<string>xxxxxx</string>
			<key>PayloadCertificateFileName</key>
			<string>iPhone.p12</string>
			<key>PayloadContent</key>
			<data>
			xxxxxx
			</data>
			<key>PayloadDescription</key>
			<string>Provides device authentication (certificate or identity).</string>
			<key>PayloadDisplayName</key>
			<string>iPhone.p12</string>
			<key>PayloadIdentifier</key>
			<string>de.xxxxxx.home.iPhone.credential</string>
			<key>PayloadOrganization</key>
			<string>home.xxxxxx.de</string>
			<key>PayloadType</key>
			<string>com.apple.security.pkcs12</string>
			<key>PayloadUUID</key>
			<string>xxxxxx</string>
			<key>PayloadVersion</key>
			<integer>1</integer>
		</dict>
	</array>
	<key>PayloadDescription</key>
	<string>OpenVPN Configuration Payload for iPhone@home.xxxxxx.de</string>
	<key>PayloadDisplayName</key>
	<string>home.xxxxxx.de OpenVPN xxxxxx@iPhone</string>
	<key>PayloadIdentifier</key>
	<string>de.xxxxxx.home.iPhone</string>
	<key>PayloadOrganization</key>
	<string>home.xxxxxx.de</string>
	<key>PayloadRemovalDisallowed</key>
	<false/>
	<key>PayloadType</key>
	<string>Configuration</string>
	<key>PayloadUUID</key>
	<string>xxxxxx</string>
	<key>PayloadVersion</key>
	<integer>1</integer>
</dict>
</plist>
```

**Comment by iphoting** [comment]
Hi, 

Please post your server logs.

**Comment by Joyphor** [comment]
My tired brain thought there wouldn't be any connection to the raspberry pi, so I didn't thought to view the logs -.-

```
Mar 14 11:03:57 NetServer ovpn-server[342]: TLS Error: tls-crypt unwrapping failed from [AF_INET]xxxxxx:25285
Mar 14 11:03:58 NetServer ovpn-server[342]: tls-crypt unwrap error: packet authentication failed
```
this loops


so I added the ta.key to the command, but no change...

**Comment by iphoting** [comment]
Looks like there's a new functionality introduced in v2.4 that this tool does not yet support: `tls-crypt`. Currently, it assumes that `ta.key` file supplied is only for the `tls-auth` option.

For a quick fix, find the `<tls-auth>...</tls-auth>` tags and substitute it with `<tls-crypt>...</tls-crypt>`. Then, delete the `<key>key-direction</key>` key and its corresponding `<string>` argument.

Let me know if this works.

**Comment by Joyphor** [comment]
where do I find the `<tls-auth>...</tls-auth>` tags?

**Comment by iphoting** [comment]
> so I added the ta.key to the command, but no change...

The iPhone.mobileconfig file output with the command quoted above.

**Comment by Joyphor** [comment]
I changed the `<key>tls-auth</key>` to `<key>tls-crypt</key>` and deleted `<key>key-direction</key>` and its corresponding `<string>` argument...

**now it's working!!**

Thank you for your fast support :-)

**Comment by iphoting** [comment]
I’ll leave this issue open until the tool supports `tls-crypt`. 

Thanks for reporting!

**Comment by Joyphor** [comment]
I saw a looping error message in the openvpn log...
This time I got the error only once right at the beginning of the connection:

```
Mar 14 20:12:37 NetServer ovpn-server[1206]: xxx.xxx.xxx.xxx:5967 tls-crypt unwrap error: bad packet ID (may be a replay): [ #1 / time = (1521054754) Wed Mar 14 20:12:34 2018 ] -- see the man page entry for --no-replay and --replay-window for more info or silence this warning with --mute-replay-warnings
Mar 14 20:12:37 NetServer ovpn-server[1206]: xxx.xxx.xxx.xxx:5967 tls-crypt unwrap error: packet replay
Mar 14 20:12:37 NetServer ovpn-server[1206]: xxx.xxx.xxx.xxx:5967 TLS Error: tls-crypt unwrapping failed from [AF_INET]xxx.xxx.xxx.xxx:5967
```

**Comment by iphoting** [comment]
If it (tls-crypt unwrap error) happens sporadically, it shouldn't be too much of a concern.

**Comment by Joyphor** [comment]
well...

```
Mar 15 18:56:07 NetServer ovpn-server[380]: VanessaiPhone7/xxx.xxx.xxx.xxx:18807 tls-crypt unwrap error: packet replay
Mar 15 18:56:07 NetServer ovpn-server[380]: VanessaiPhone7/xxx.xxx.xxx.xxx:18807 TLS Error: tls-crypt unwrapping failed from [AF_INET]xxx.xxx.xxx.xxx:18807
Mar 15 18:56:08 NetServer ovpn-server[380]: VanessaiPhone7/xxx.xxx.xxx.xxx:18807 tls-crypt unwrap error: bad packet ID (may be a replay): [ #9 / time = (1521136542) Thu Mar 15 18:55:42 2018 ] -- see the man page entry for --no-replay and --replay-window for more info or silence this warning with --mute-replay-warnings
Mar 15 18:56:08 NetServer ovpn-server[380]: VanessaiPhone7/xxx.xxx.xxx.xxx:18807 tls-crypt unwrap error: packet replay
Mar 15 18:56:08 NetServer ovpn-server[380]: VanessaiPhone7/xxx.xxx.xxx.xxx:18807 TLS Error: tls-crypt unwrapping failed from [AF_INET]xxx.xxx.xxx.xxx:18807
Mar 15 18:56:10 NetServer ovpn-server[380]: VanessaiPhone7/xxx.xxx.xxx.xxx:18807 tls-crypt unwrap error: bad packet ID (may be a replay): [ #9 / time = (1521136542) Thu Mar 15 18:55:42 2018 ] -- see the man page entry for --no-replay and --replay-window for more info or silence this warning with --mute-replay-warnings
Mar 15 18:56:10 NetServer ovpn-server[380]: VanessaiPhone7/xxx.xxx.xxx.xxx:18807 tls-crypt unwrap error: packet replay
Mar 15 18:56:10 NetServer ovpn-server[380]: VanessaiPhone7/xxx.xxx.xxx.xxx:18807 TLS Error: tls-crypt unwrapping failed from [AF_INET]xxx.xxx.xxx.xxx:18807
Mar 15 18:56:11 NetServer ovpn-server[380]: VanessaiPhone7/xxx.xxx.xxx.xxx:18807 tls-crypt unwrap error: bad packet ID (may be a replay): [ #9 / time = (1521136542) Thu Mar 15 18:55:42 2018 ] -- see the man page entry for --no-replay and --replay-window for more info or silence this warning with --mute-replay-warnings
Mar 15 18:56:11 NetServer ovpn-server[380]: VanessaiPhone7/xxx.xxx.xxx.xxx:18807 tls-crypt unwrap error: packet replay
Mar 15 18:56:11 NetServer ovpn-server[380]: VanessaiPhone7/xxx.xxx.xxx.xxx:18807 TLS Error: tls-crypt unwrapping failed from [AF_INET]xxx.xxx.xxx.xxx:18807
Mar 15 18:56:12 NetServer ovpn-server[380]: VanessaiPhone7/xxx.xxx.xxx.xxx:18807 tls-crypt unwrap error: bad packet ID (may be a replay): [ #9 / time = (1521136542) Thu Mar 15 18:55:42 2018 ] -- see the man page entry for --no-replay and --replay-window for more info or silence this warning with --mute-replay-warnings
Mar 15 18:56:12 NetServer ovpn-server[380]: VanessaiPhone7/xxx.xxx.xxx.xxx:18807 tls-crypt unwrap error: packet replay
Mar 15 18:56:12 NetServer ovpn-server[380]: VanessaiPhone7/xxx.xxx.xxx.xxx:18807 TLS Error: tls-crypt unwrapping failed from [AF_INET]xxx.xxx.xxx.xxx:18807
Mar 15 18:56:13 NetServer ovpn-server[380]: VanessaiPhone7/xxx.xxx.xxx.xxx:18807 tls-crypt unwrap error: bad packet ID (may be a replay): [ #9 / time = (1521136542) Thu Mar 15 18:55:42 2018 ] -- see the man page entry for --no-replay and --replay-window for more info or silence this warning with --mute-replay-warnings
Mar 15 18:56:13 NetServer ovpn-server[380]: VanessaiPhone7/xxx.xxx.xxx.xxx:18807 tls-crypt unwrap error: packet replay
Mar 15 18:56:13 NetServer ovpn-server[380]: VanessaiPhone7/xxx.xxx.xxx.xxx:18807 TLS Error: tls-crypt unwrapping failed from [AF_INET]xxx.xxx.xxx.xxx:18807
Mar 15 18:56:14 NetServer ovpn-server[380]: VanessaiPhone7/xxx.xxx.xxx.xxx:18807 tls-crypt unwrap error: bad packet ID (may be a replay): [ #9 / time = (1521136542) Thu Mar 15 18:55:42 2018 ] -- see the man page entry for --no-replay and --replay-window for more info or silence this warning with --mute-replay-warnings
Mar 15 18:56:14 NetServer ovpn-server[380]: VanessaiPhone7/xxx.xxx.xxx.xxx:18807 tls-crypt unwrap error: packet replay
Mar 15 18:56:14 NetServer ovpn-server[380]: VanessaiPhone7/xxx.xxx.xxx.xxx:18807 TLS Error: tls-crypt unwrapping failed from [AF_INET]xxx.xxx.xxx.xxx:18807
Mar 15 18:56:15 NetServer ovpn-server[380]: VanessaiPhone7/xxx.xxx.xxx.xxx:18807 tls-crypt unwrap error: bad packet ID (may be a replay): [ #9 / time = (1521136542) Thu Mar 15 18:55:42 2018 ] -- see the man page entry for --no-replay and --replay-window for more info or silence this warning with --mute-replay-warnings
Mar 15 18:56:15 NetServer ovpn-server[380]: VanessaiPhone7/xxx.xxx.xxx.xxx:18807 tls-crypt unwrap error: packet replay
Mar 15 18:56:15 NetServer ovpn-server[380]: VanessaiPhone7/xxx.xxx.xxx.xxx:18807 TLS Error: tls-crypt unwrapping failed from [AF_INET]xxx.xxx.xxx.xxx:18807
Mar 15 18:56:16 NetServer ovpn-server[380]: VanessaiPhone7/xxx.xxx.xxx.xxx:18807 tls-crypt unwrap error: bad packet ID (may be a replay): [ #9 / time = (1521136542) Thu Mar 15 18:55:42 2018 ] -- see the man page entry for --no-replay and --replay-window for more info or silence this warning with --mute-replay-warnings
Mar 15 18:56:16 NetServer ovpn-server[380]: VanessaiPhone7/xxx.xxx.xxx.xxx:18807 tls-crypt unwrap error: packet replay
Mar 15 18:56:16 NetServer ovpn-server[380]: VanessaiPhone7/xxx.xxx.xxx.xxx:18807 TLS Error: tls-crypt unwrapping failed from [AF_INET]xxx.xxx.xxx.xxx:18807
Mar 15 18:56:17 NetServer ovpn-server[380]: VanessaiPhone7/xxx.xxx.xxx.xxx:18807 tls-crypt unwrap error: bad packet ID (may be a replay): [ #9 / time = (1521136542) Thu Mar 15 18:55:42 2018 ] -- see the man page entry for --no-replay and --replay-window for more info or silence this warning with --mute-replay-warnings
Mar 15 18:56:17 NetServer ovpn-server[380]: VanessaiPhone7/xxx.xxx.xxx.xxx:18807 tls-crypt unwrap error: packet replay
Mar 15 18:56:17 NetServer ovpn-server[380]: VanessaiPhone7/xxx.xxx.xxx.xxx:18807 TLS Error: tls-crypt unwrapping failed from [AF_INET]xxx.xxx.xxx.xxx:18807
```

**Comment by joubin** [comment]
Same here. I would also like to note that I do not have a section for auth(<tls-auth>...</tls-auth>) in my mobileconfig. But the error is the same.

**Comment by iphoting** [comment]
Resolved in #23 and 21435de7181e2eb3f9d98a1133e502055165d90e.
